janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/fetcher/LabelStepBatchFetcher.java

```diff
@@ -0,0 +1,48 @@
+// Copyright 2023 JanusGraph Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize.step.fetcher;
+
+import com.google.common.collect.Iterables;
+import org.janusgraph.core.JanusGraphMultiVertexQuery;
+import org.janusgraph.core.JanusGraphVertex;
+import org.janusgraph.graphdb.query.vertex.BasicVertexCentricQueryUtil;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class LabelStepBatchFetcher extends MultiQueriableStepBatchFetcher<String>{
+
+    private final FetchQueryBuildFunction fetchQueryBuildFunction;
+
+    public LabelStepBatchFetcher(FetchQueryBuildFunction fetchQueryBuildFunction, int batchSize) {
+        super(batchSize);
+        this.fetchQueryBuildFunction = fetchQueryBuildFunction;
+    }
+
+    @Override
+    protected Map<JanusGraphVertex, String> makeQueryAndExecute(JanusGraphMultiVertexQuery multiQuery) {
+        multiQuery = fetchQueryBuildFunction.makeQuery(multiQuery);
+        Map<JanusGraphVertex, Iterable<JanusGraphVertex>> labelsBatch = multiQuery.vertices();
+        Map<JanusGraphVertex, String> result = new HashMap<>(labelsBatch.size());
+        for(Map.Entry<JanusGraphVertex, Iterable<JanusGraphVertex>> labelEntry : labelsBatch.entrySet()){
```

Comment:
We can leverage the fact that the old and new hashmaps contain exactly the same keys:

```java
Map<JanusGraphVertex, Iterable<JanusGraphVertex>> labelsBatch = multiQuery.vertices();

Map<JanusGraphVertex, String> result = new HashMap<>(labelsBatch);

result.replaceAll((key, value) -> BasicVertexCentricQueryUtil.castToVertexLabel(Iterables.getOnlyElement(value,null)).name());
```

Reply:
> We can leverage the fact that the old and new hashmaps contain exactly the same keys:
> 
> ```java
> Map<JanusGraphVertex, Iterable<JanusGraphVertex>> labelsBatch = multiQuery.vertices();
> 
> Map<JanusGraphVertex, String> result = new HashMap<>(labelsBatch);
> 
> result.replaceAll((key, value) -> BasicVertexCentricQueryUtil.castToVertexLabel(Iterables.getOnlyElement(value,null)).name());
> ```

I'm not fully getting this comment :sweat_smile: 
Could be that I'm missing something but you are trying to copy data from one map into another map with incompatible value types. This doesn't compile.
Moreover, even if it did compile and we would copy that data (let's assume we change Parameterized map types to `Object`. I.e. `Map<JanusGraphVertex, Object>` ) - I'm not getting why two operations (copy map entries and then replacing each entry value) is more beneficial than just copying already converted values? Feels like with your approach we would need to iterate map entries 2 times, but with the current approach we iterate only once.

Could be I'm missing something, so, please, correct me if I'm wrong.

Reply:
Oh yeah, the types are not compatible so it would be hacky to make it compile. Never mind - sorry I didn't try compiling this code by myself.

My point was since the keys remain the same, it seems unnecessary to hash the keys again - when you do `result.put(labelEntry.getKey(), ...)`, it does a hash operation to find the bucket in the hash table. So, if we could somehow just clone the map and replace the values, it is theoretically more efficient.